src/app/check-in/check-in.component.ts

```diff
@@ -13,96 +13,96 @@ import { ThsDataService } from '../services/ths-data.service';
+      console.log(this.lastName);
+      console.log(this.email);
+      this.router.navigateByUrl('/appointments');
+      console.log('log in with ' + this.patientId);
+    } else {
+      this.authenticationFlag = false;
+      this.patientId = '';
+      console.log('failed log in ' + this.patientId);
+    }
+  }
 
-   constructor(private router: Router, private tsDataService: TsScreenerDataService, private tfiDataService: TfiDataService, private thsDataService: ThsDataService) {};
+  isCredentialsValid() : boolean {
+    return this.isEmailValid() && this.isPatientIdValid();
+  }
 
-   public numRearrange(num) {
-      let fst = Number(num[0]);
-      let snd = Number(num[1]);
-      let thd = Number(num[2]);
-      let frt = Number(num[3]);
-      if (frt < 9) {
-        frt += 1;
-      } else { frt -= 5; }
+  isEmailValid() : boolean {
+    let numAtSymbols : number = 0;
 
-      if (thd < 5) {
-        thd += 4;
-      } else { thd -= 2; }
+    // Email is an optional field the user does not need
+    // to fill out. However, should they choose to do so, it
+    // should follow a proper format.
+    if(this.email.length === 0)
+      return true;
 
-      if (snd < 7) {
-        snd += 2;
-      } else { snd -= 3; }
+    let i: number;
+    for(i = 0; i < this.email.length; i++) {
+      if(this.email.charAt(i) === '@')
+        numAtSymbols++;
+    }
 
-      if (fst < 7) {
-        fst += 2;
-      } else { fst -= 2; }
+    if(numAtSymbols !== 1)
+      return false;
 
-      let result1 = thd.toString();
-      let result2 = fst.toString();
-      let result3 = frt.toString();
-      let result4 = snd.toString();
+    // Credit for this here: https://tylermcginnis.com/validate-email-address-javascript/
+    // Doesn't seem to handle multiple @ symbols however.
+    let regexp = new RegExp('^[^\s@]+@[^\s@]+\.[^\s@]+$');
 
-      return result1 + result2 + result3 + result4;
-   }
-   public idGenerator() {
-     let hash = 0;
-     let first = this.firstName;
-     let last = this.lastName;
-     // name = this.firstName.charCodeAt(0) + this.lastName.charCodeAt(0);
-     if (this.firstName == null || this.lastName == null || this.lastFourSS == null) {
-       return hash;
-     }
-     // for (let i = 0; i < nameLength; i++) {
-     //   hash += Math.pow(name.charCodeAt(i) * 31, nameLength - i);
-     // }
-     let result = first.substr(0, 3) + last.substr(0, 3) + this.numRearrange(this.lastFourSS);
-     return result;
-   }
+    return regexp.test(this.email);
+  }
 
-   /**
-    * This function will be call when the "check in" button is pressed.
-    * It evaluates the value is typed in the input area; if the value has the length of 6,
-    * then the page will be route to the "audiologist" page;
-    * if the value of length 4, then the page will be route to the "appointment" page.
-    * If the value has different length, then it will be informed as incorrect check in.
-    */
+  isPatientIdValid() : boolean {
```

Comment:
You can actually use Regular expression to your advantage here: `isPatientIdValid` and `isPatientIdNumber` is all equivalent to the following code:

```  javascript
isPatientIdValid(): boolean {
    const PATIENT_ID_LENGTH = 4;
    let reg = new RegExp('^\\d{'+ PATIENT_ID_LENGTH +'}$');
    return (reg.test(this.patientId)) ? true : false;
  }
```